### PR TITLE
adding possibility to add SMTP configuration

### PR DESCRIPTION
**Pull request by tomrutsaert**
Adding possibility to add SMTP Server configuration.

### CODE REVIEW

File: keycloak/realm.go [hunk 1]
```diff
@@ -92,6 +110,14 @@ func (keycloakClient *KeycloakClient) ValidateRealm(realm *Realm) error {
 		return err
 	}
 
+	if (SmtpServer{}) != realm.SmtpServer && realm.SmtpServer.Host == "" {
```
Comment (BernhardBerbuir) [inline keycloak/realm.go]: is there a reason why this is not checked by Terraform?
This attribute is configured as `Optional: true,`
=> could it be configured as`Required: true,`?

Comment (tomrutsaert) [reply keycloak/realm.go]: Well it is an optional field in the sense that you do need to set smtp server, but the moment you set one field of the smtp server, this is one of two required fields. (host and from).
For that reason, I do not think I can make the field required, can I?
Not really a Golang orTerraform expert, But I will take a look tomorrow, what it gives when I make this field required but not set the smtpserver

Comment (tomrutsaert) [reply keycloak/realm.go]: changed it

File: provider/resource_keycloak_realm.go [hunk 2]
```diff
@@ -73,6 +73,70 @@ func resourceKeycloakRealm() *schema.Resource {
...
+						},
+						"user": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"password": {
+							Type:      schema.TypeString,
+							Optional:  true,
+							Sensitive: true,
+							DiffSuppressFunc: func(_, remoteBindCredential, _ string, _ *schema.ResourceData) bool {
```
Comment (mrparkers) [inline provider/resource_keycloak_realm.go]: Could you rename `remoteBindCredential`?  I think this was copied from the ldap user federation resource where this particular attribute was called `bind_credential`.

File: provider/resource_keycloak_realm_test.go [hunk 3]
```diff
@@ -429,6 +488,76 @@ resource "keycloak_realm" "realm" {
...
+	return fmt.Sprintf(`
+resource "keycloak_realm" "realm" {
+	realm        = "%s"
+	enabled      = true
+	display_name = "%s"
+	smtp_server {
+      host = "%s"
+	  port = 25
+	  from_display_name = "Tom"
+      from = "%s"
```
Comment (mrparkers) [inline provider/resource_keycloak_realm_test.go]: would you be able to fix the indentation for these SMTP config blocks?

File: provider/resource_keycloak_realm_test.go [hunk 4]
```diff
@@ -86,6 +86,46 @@ func TestAccKeycloakRealm_import(t *testing.T) {
...
+				Config:      testKeycloakRealm_WithSmtpServerWithoutHost(realm, "My Host"),
+				ExpectError: regexp.MustCompile("config is invalid: Missing required argument: The argument \"host\" is required, but no definition was found."),
+			},
+			{
+				Config:      testKeycloakRealm_WithSmtpServerWithoutFrom(realm, "myhost.com"),
+				ExpectError: regexp.MustCompile("config is invalid: Missing required argument: The argument \"from\" is required, but no definition was found."),
+			},
+		},
+	})
+}
```
Comment (mrparkers) [inline provider/resource_keycloak_realm_test.go]: thanks for adding these tests!  would you be able to add one more that creates a realm with some SMTP settings and updates those settings in a second test step?

File: keycloak/realm.go (hunk 1 above)
Comment (mrparkers) [reply keycloak/realm.go]: thanks for reviewing this @BernhardBerbuir!  I would have left the same comment :smile:

File: provider/resource_keycloak_realm.go (hunk 2 above)
Comment (tomrutsaert) [reply provider/resource_keycloak_realm.go]: ok

File: provider/resource_keycloak_realm_test.go (hunk 3 above)
Comment (tomrutsaert) [reply provider/resource_keycloak_realm_test.go]: strange, I do not have this weird indentation, I will copy and paste the whole block again

File: provider/resource_keycloak_realm_test.go (hunk 4 above)
Comment (tomrutsaert) [reply provider/resource_keycloak_realm_test.go]: Yes, I will do that.

**Comment by tomrutsaert** [comment]
All request changes were added + I have updated the branch to latest version of master

**Comment by mrparkers** [comment]
@tomrutsaert I just noticed this after merging, but what would you think of handling SMTP auth like this instead?

```hcl
resource "keycloak_realm" "test" {
  realm        = "test"

  smtp_server {
    host = "mysmtphost.com"
    from = "tom@myhost.com"
    auth {
      user = "tom"
      password = "tom"
    }
  }
}
```

If you're okay with this, I'll push up a quick fix and cut a release for you within the next few days.

**Comment by tomrutsaert** [comment]
Seems like a good idea, go ahead

**Pull request merged**